---
title: CA2000:丢失范围之前释放对象
ms.date: 05/14/2019
ms.topic: reference
f1_keywords:
- CA2000
- Dispose objects before losing scope
- DisposeObjectsBeforeLosingScope
helpviewer_keywords:
- CA2000
- DisposeObjectsBeforeLosingScope
ms.assetid: 0c3d7d8d-b94d-46e8-aa4c-38df632c1463
author: gewarren
ms.author: gewarren
manager: jillfra
dev_langs:
- CSharp
- VB
ms.workload:
- multiple
ms.openlocfilehash: 7a498a01741b86c16a52f790489dc8ce62aad06c
ms.sourcegitcommit: 3e94d9fb6dc56fa8b23fbacd5d11cf8d6e7e18f1
ms.translationtype: MT
ms.contentlocale: zh-CN
ms.lasthandoff: 10/10/2019
ms.locfileid: "72256779"
---
# <a name="ca2000-dispose-objects-before-losing-scope"></a>CA2000:丢失范围之前释放对象

|||
|-|-|
|TypeName|DisposeObjectsBeforeLosingScope|
|CheckId|CA2000|
|类别|Microsoft.Reliability|
|重大更改|不间断|

## <a name="cause"></a>原因

创建了 @no__t 类型的本地对象，但在对该对象的所有引用超出范围之前，不会释放该对象。

## <a name="rule-description"></a>规则说明

如果在对某个可释放对象的所有引用超出范围之前未显式释放该对象，则当垃圾回收器运行该对象的终结器时，将在某个不确定的时间释放该对象。 由于可能发生异常事件，将阻止对象的终结器运行，因此应改为显式释放对象。

### <a name="special-cases"></a>特殊情况

即使未释放对象，也不会为以下类型的本地对象触发规则 CA2000：

- <xref:System.IO.Stream?displayProperty=nameWithType>
- <xref:System.IO.TextReader?displayProperty=nameWithType>
- <xref:System.IO.TextWriter?displayProperty=nameWithType>
- <xref:System.Resources.IResourceReader?displayProperty=nameWithType>

将其中一个类型的对象传递给构造函数，然后将其分配给一个字段，表示将一个*处置所有权转移*到新构造的类型。 也就是说，新构造的类型现在负责释放对象。 如果你的代码将其中一个类型的对象传递给构造函数，则即使在对对象的所有引用超出范围之前未释放该对象，也不会发生规则 CA2000 冲突。

## <a name="how-to-fix-violations"></a>如何解决冲突

若要修复与此规则的冲突，请在对象的所有引用超出范围之前，对该对象调用 <xref:System.IDisposable.Dispose%2A>。

您可以使用[`using` 语句](/dotnet/csharp/language-reference/keywords/using-statement)（Visual Basic 中[@no__t](/dotnet/visual-basic/language-reference/statements/using-statement) ）来包装实现 <xref:System.IDisposable> 的对象。 以这种方式包装的对象将自动在 `using` 块结束时释放。 但是，以下情况不应使用 `using` 语句来处理：

- 若要返回可释放对象，对象必须在 @no__t 块外的 `try/finally` 块中构造。

- 不要在 `using` 语句的构造函数中初始化可释放对象的成员。

- 如果仅由一个异常处理程序保护的构造函数嵌套在[`using` 语句的获取部分](/dotnet/csharp/language-reference/language-specification/statements#the-using-statement)中，则外部构造函数中的失败会导致不会关闭嵌套构造函数所创建的对象。 在下面的示例中，<xref:System.IO.StreamReader> 构造函数中的失败会导致不会关闭 @no__t 1 对象。 在这种情况下，CA2000 标志违反规则。

   ```csharp
   using (StreamReader sr = new StreamReader(new FileStream("C:\myfile.txt", FileMode.Create)))
   { ... }
   ```

- 动态对象应使用影子对象来实现 <xref:System.IDisposable> 对象的释放模式。

## <a name="when-to-suppress-warnings"></a>何时禁止显示警告

请勿禁止显示此规则发出的警告，除非：

- 你在对象上调用了一个方法，该方法调用 `Dispose`，如 <xref:System.IO.Stream.Close%2A>
- 引发警告的方法将返回包装对象的 @no__t 0 对象
- 分配方法没有释放所有权;也就是说，释放对象的责任将传输到在方法中创建并返回给调用方的另一个对象或包装器

## <a name="related-rules"></a>相关规则

- [CA2213：应释放可释放的字段](../code-quality/ca2213-disposable-fields-should-be-disposed.md)
- [CA2202：不要多次释放对象 @ no__t-0

## <a name="example"></a>示例

如果要实现一个返回可释放对象的方法，请在没有 catch 块的情况下使用 try/finally 块来确保释放对象。 通过使用 try/finally 块，你允许在错误点引发异常，并确保对象已释放。

在 OpenPort1 方法中，打开 ISerializable 对象 SerialPort 或对 SomeMethod 的调用可能会失败。 此实现将引发 CA2000 警告。

在 OpenPort2 方法中，将声明两个 SerialPort 对象并将其设置为 null：

- `tempPort`，用于测试方法操作是否成功。

- `port`，用于方法的返回值。

@No__t-0 在 @no__t 块中构造和打开，而任何其他必需的工作则在同一 @no__t 2 块中执行。 在 `try` 块的末尾，打开的端口分配给将返回的 @no__t 对象，@no__t 2 的对象设置为 `null`。

@No__t-0 块检查 `tempPort` 的值。 如果它不为 null，则方法中的操作失败，并 `tempPort` 关闭以确保释放所有资源。 如果方法的操作成功，则返回的端口对象将包含打开的 SerialPort 对象; 如果操作失败，则将为 null。

```csharp
public SerialPort OpenPort1(string portName)
{
   SerialPort port = new SerialPort(portName);
   port.Open();  //CA2000 fires because this might throw
   SomeMethod(); //Other method operations can fail
   return port;
}

public SerialPort OpenPort2(string portName)
{
   SerialPort tempPort = null;
   SerialPort port = null;
   try
   {
      tempPort = new SerialPort(portName);
      tempPort.Open();
      SomeMethod();
      //Add any other methods above this line
      port = tempPort;
      tempPort = null;

   }
   finally
   {
      if (tempPort != null)
      {
         tempPort.Close();
      }
   }
   return port;
}
```

```vb
Public Function OpenPort1(ByVal PortName As String) As SerialPort

   Dim port As New SerialPort(PortName)
   port.Open()    'CA2000 fires because this might throw
   SomeMethod()   'Other method operations can fail
   Return port

End Function

Public Function OpenPort2(ByVal PortName As String) As SerialPort

   Dim tempPort As SerialPort = Nothing
   Dim port As SerialPort = Nothing

   Try
      tempPort = New SerialPort(PortName)
      tempPort.Open()
      SomeMethod()
      'Add any other methods above this line
      port = tempPort
      tempPort = Nothing

   Finally
      If Not tempPort Is Nothing Then
         tempPort.Close()
      End If

   End Try

   Return port

End Function
```

## <a name="example"></a>示例

默认情况下，Visual Basic 编译器会检查溢出情况下的所有算术运算符。 因此，任何 Visual Basic 算术运算都可能引发 @no__t。 这可能会导致 CA2000 等规则发生意外冲突。 例如，以下 CreateReader1 函数将产生 CA2000 冲突，因为 Visual Basic 编译器发出的溢出检查指令可能会引发导致 StreamReader 无法释放的异常。

若要解决此问题，可以在项目中禁用 Visual Basic 编译器发出溢出检查，也可以修改代码，如以下 CreateReader2 函数所示。

若要禁用溢出检查，请在解决方案资源管理器中右键单击项目名称，然后单击 "**属性**"。 单击 "**编译**"，单击 "**高级编译选项**"，然后选中 "**删除整数溢出检查**"。

[!code-vb[FxCop.Reliability.CA2000.DisposeObjectsBeforeLosingScope#1](../code-quality/codesnippet/VisualBasic/ca2000-dispose-objects-before-losing-scope-vboverflow_1.vb)]

## <a name="see-also"></a>请参阅

- <xref:System.IDisposable>
- [Dispose 模式](/dotnet/standard/design-guidelines/dispose-pattern)